### PR TITLE
[REEF-1019] Implement .NET REEF client using YARN REST APIs

### PR DESCRIPTION
**Pull request by anupam128**
This addressed the issue by
 * Implementing IREEFClient using YARN REST submission
 * Creating a mode in HelloREEF to exercise this code

JIRA:
 [REEF-1019](https://issues.apache.org/jira/browse/REEF-1019)

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote.Parameters;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
+    /// Temporary client for developing and testing .NET job submission E2E.
+    /// When REEF-70 is completed YARNREEFClient should be either merged or
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: Mark this with a `// TODO[REEF-70]`

File: lang/cs/Org.Apache.REEF.Client.Tests/WindowsYarnJobCommandProviderTests.cs [hunk 2]
```diff
@@ -50,9 +50,7 @@ public class WindowsYarnJobCommandProviderTests
             "%HADOOP_HOME%/share/hadoop/hdfs/*",
             "%HADOOP_HOME%/share/hadoop/hdfs/lib/*",
             "%HADOOP_HOME%/share/hadoop/mapreduce/*",
-            "%HADOOP_HOME%/share/hadoop/mapreduce/lib/*",
-            "reef/local/*",
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client.Tests/WindowsYarnJobCommandProviderTests.cs]: Why are these reef folders gone?

File: lang/cs/Org.Apache.REEF.Client.Tests/YarnClientTests.cs [hunk 3]
```diff
@@ -231,7 +229,7 @@ public class YarnClientTests
                     },
                     LocalResources = new LocalResources
                     {
-                        Entries = new List<YARN.RestClient.DataModel.KeyValuePair<string, LocalResourcesValue>>
+                        Entries = new System.Collections.Generic.List<YARN.RestClient.DataModel.KeyValuePair<string, LocalResourcesValue>>
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client.Tests/YarnClientTests.cs]: What was the change here?

File: lang/cs/Org.Apache.REEF.Client/YARN/IJobResourceUploader.cs [hunk 4]
```diff
@@ -26,7 +26,8 @@ public interface IJobResourceUploader
         /// Upload archived local driver folder to DFS destination path.
         /// </summary>
         /// <param name="driverLocalFolderPath">Local folder where REEF application resources are staged</param>
+        /// <param name="jobSubmissionDirectory">Remote directory path where we will upload resources</param>
         /// <returns>Path, modification time and size of uploaded file as JobResource</returns>
-        JobResource UploadJobResource(string driverLocalFolderPath);
+        JobResource UploadJobResource(string driverLocalFolderPath, string jobSubmissionDirectory);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobResourceUploader.cs]: Do we need to expose the directory here? The original idea was to keep that information in the configuration of the uploader.

Either way, this is a `public` interface and if possible, we should keep the old signature and maybe `[Obsolete]` it.

File: lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs [hunk 5]
```diff
@@ -15,34 +15,39 @@
...
 using Org.Apache.REEF.Client.YARN.Parameters;
 using Org.Apache.REEF.Common.Files;
 using Org.Apache.REEF.Tang.Annotations;
 
 namespace Org.Apache.REEF.Client.Yarn
 {
+    /// <summary>
+    /// Provides path to job submission directory.
+    /// </summary>
     public class JobSubmissionDirectoryProvider : IJobSubmissionDirectoryProvider
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs]: sealed?

File: lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs [hunk 6]
```diff
@@ -29,6 +29,6 @@ public interface IJobSubmissionDirectoryProvider
         /// Returns path to job submission directory in DFS.
         /// </summary>
         /// <returns></returns>
-        string GetJobSubmissionRemoteDirectory();
+        string GetJobSubmissionRemoteDirectory(string applicationId);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs]: Similar question as above: Do we absolutely need the new parameter and if we do, can we keep the old method as well?

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnCommandLineEnvironment.cs [hunk 7]
```diff
@@ -30,6 +30,6 @@ internal interface IYarnCommandLineEnvironment
         /// Returns the class path returned by `yarn classpath`.
         /// </summary>
         /// <returns>The class path returned by `yarn classpath`.</returns>
-        IList<string> GetYarnClasspathList();
+        IReadOnlyList<string> GetYarnClasspathList();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnCommandLineEnvironment.cs]: I like the change, but it is a breaking change, right?

File: lang/cs/Org.Apache.REEF.Client/YARN/LegacyJobResourceUploader.cs [hunk 8]
```diff
@@ -40,26 +40,26 @@ internal sealed class LegacyJobResourceUploader : IJobResourceUploader
...
 
         private readonly IJavaClientLauncher _javaLauncher;
-        private readonly IJobSubmissionDirectoryProvider _jobSubmissionDirectoryProvider;
         private readonly IResourceArchiveFileGenerator _resourceArchiveFileGenerator;
         private readonly IFile _file;
 
         [Inject]
-        private LegacyJobResourceUploader(IJavaClientLauncher javaLauncher,
-            IJobSubmissionDirectoryProvider jobSubmissionDirectoryProvider,
+        private LegacyJobResourceUploader(
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/LegacyJobResourceUploader.cs]: This is only executed when submitting from the Cluster right?

File: lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs [hunk 9]
```diff
@@ -15,34 +15,39 @@
 // specific language governing permissions and limitations
 // under the License.
 
-using System;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs]: The changes in this file are all cosmetic, right? I like them, but wonder whether they should go to a separate PR.

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+using Org.Apache.REEF.Driver.Bridge;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote.Parameters;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: Sorry, I'm a bit confused by the comment here, along with all the items we have for a client that doesn't require Java. Could you clarify the differences between the following items?
[REEF-227](https://issues.apache.org/jira/browse/REEF-227), [REEF-189](https://issues.apache.org/jira/browse/REEF-189), [REEF-70](https://issues.apache.org/jira/browse/REEF-70), [REEF-908](https://issues.apache.org/jira/browse/REEF-908), and [REEF-1019](https://issues.apache.org/jira/browse/REEF-1019). It seems like there's a lot of overlap between the descriptions.

**Comment by markusweimer** [comment]
I've done a sruface-level pass. Overall looks good, but I left some minor nitpicks about unrelated changes and potential breakage of `public`, yet `[Unstable]` APIs.

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+            return submitApplication;
+        }
+
+        /// <summary>
+        /// Creates the temporary directory to hold the job submission.
+        /// </summary>
+        /// <returns>The path to the folder created.</returns>
+        private string CreateDriverFolder(string jobId, string appId)
+        {
+            return Path.GetFullPath(Path.Combine(Path.GetTempPath(), string.Join("-", "reef", jobId, appId)));
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: There was a thread about discouraging system APIs @dev. See http://mail-archives.apache.org/mod_mbox/reef-dev/201512.mbox/browser. I'm not sure if it's applicable here though. @markusweimer perhaps you can provide some more insight on this.

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+            var submissionReq = CreateApplicationSubmissionRequest(jobSubmission,
+                applicationId,
+                maxApplicationSubmissions,
+                jobResource);
+            var submittedApplication = _yarnRMClient.SubmitApplicationAsync(submissionReq).GetAwaiter().GetResult();
+            Log.Log(Level.Info, @"Submitted application {0}", submittedApplication.Id);
+        }
+
+        public IJobSubmissionResult SubmitAndGetJobStatus(IJobSubmission jobSubmission)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: Will this be implemented? If not, maybe `NotSupportedException` is a better choice.

File: lang/cs/Org.Apache.REEF.Client.Tests/WindowsYarnJobCommandProviderTests.cs (hunk 2 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client.Tests/WindowsYarnJobCommandProviderTests.cs]: They are not gone. They are still there in the `expectedCommand`. They are just explicitly added as classpath. Everything except the REEF folders come from `IYarnCommandLineEnvironment`. This is just setting up that portion.

File: lang/cs/Org.Apache.REEF.Client/YARN/IJobResourceUploader.cs (hunk 4 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobResourceUploader.cs]: One instance of uploader should be able to upload in different locations, if need by. Earlier we thought we would generate `jobSubmissionDIrectory` in `IJobResourceUploader` which was a bad idea in hindsight, creating hidden coupling between `IJobSubmissionDirectoryProvider` and `IJobResourceUploader`.
I would strongly prefer not to go the `Obsolete` route here as it is not possible for anyone to take dependency on this API right now. It is just added work for no good reason. Thoughts?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs (hunk 11 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: This is running on the client, which isn't generally managed by a resource manager. Hence, this should be fine.

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnCommandLineEnvironment.cs (hunk 7 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnCommandLineEnvironment.cs]: @markusweimer Yes, this is indeed is a breaking change. I recently added this interface (within this release cycle). I would much rather revert this change than go through `Obsolete` song and dance. 
This being mutable gave me a lot of grief in debugging a test failure and is definitely a possible source of error. Let me know what you want.

File: lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs (hunk 9 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs]: I can revert. Separate PR never ends up happening...

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs (hunk 11 above)
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: I see, thanks!

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+            // getting wrapped in AggregateException.
+            var newApplication = _yarnRMClient.CreateNewApplicationAsync().GetAwaiter().GetResult();
+            string applicationId = newApplication.ApplicationId;
+
+            // create job submission remote path
+            string jobSubmissionDirectory =
+                _jobSubmissionDirectoryProvider.GetJobSubmissionRemoteDirectory(applicationId);
+
+            // create local driver folder.
+            string localDriverFolderPath = CreateDriverFolder(jobId, applicationId);
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: try-finally to delete the folder?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs (hunk 10 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: Valid confusion. REEF-70 is the wrong JIRA to cite here. Changing to REEF-189. We will have to do some clean up of the JIRAs.

File: lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs (hunk 6 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs]: Adding applicationId here makes it easier to debug. Not worth it to keep two methods.

File: lang/cs/Org.Apache.REEF.Client/YARN/IYarnCommandLineEnvironment.cs (hunk 7 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IYarnCommandLineEnvironment.cs]: Ah, OK. Let's go ahead with it then.

File: lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs (hunk 9 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/JobSubmissionDirectoryProvider.cs]: Maybe just call it out in the commit message will do.

File: lang/cs/Org.Apache.REEF.Client/YARN/IJobResourceUploader.cs (hunk 4 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobResourceUploader.cs]: Right, move ahead as noone can take a dependency right now.

File: lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs (hunk 6 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs]: Who calls this API?

Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/IJobSubmissionDirectoryProvider.cs]: The client (`YarnREEFDotNetClient`) during `Submit` (or similar call in the future).

File: lang/cs/Org.Apache.REEF.Client.Tests/YarnClientTests.cs (hunk 3 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client.Tests/YarnClientTests.cs]: Reverted. Some IDE quirk.

File: lang/cs/Org.Apache.REEF.Client/YARN/LegacyJobResourceUploader.cs (hunk 8 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/LegacyJobResourceUploader.cs]: Yes, this preserves the exact behavior as with current Java based submission for resource uploading.

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs (hunk 12 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: We will change this API in future.

**Comment by anupam128** [comment]
@afchung @markusweimer made changes. Please take a look.

File: lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.IO;
+
+namespace Org.Apache.REEF.Client.Common
+{
+    internal sealed class TempDirectory : IDisposable
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs]: Would be great if you could provide some documentation here.

File: lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        private readonly string _directoryPath;
+        private readonly object _lock;
+        public TempDirectory(string directoryPath)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs]: nit: new line

File: lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            _directoryPath = directoryPath;
+            _lock = new object();
+        }
+
+        public void Dispose()
+        {
+            lock (_lock)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/Common/TempDirectory.cs]: Seems a bit strange for disposable to dispose of something it doesn't create. Could the object create the folder as well? i.e. modify `_driverFolderPreparationHelper.PrepareDriverFolder()` to return a new `TempDirectory` with the root directory created in the constructor.

File: lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc [hunk 17]
```diff
@@ -49,7 +49,7 @@
         { "name": "sharedJobSubmissionParameters", "type": "AvroJobSubmissionParameters" },
         { "name": "driverMemory", "type": "int" },
         { "name": "driverRecoveryTimeout", "type": "int" },
-        { "name": "dfsJobSubmissionFolder", "type": ["null", "string"], "default" : null },
+        { "name": "dfsJobSubmissionFolder", "type": "string", "default": "NULL" },
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc]: Why are we changing this to a string literal "NULL"?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote.Parameters;
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
+    /// Temporary client for developing and testing .NET job submission E2E.
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: This is a temporary client, as in this class will be replaced? Or just modified? If it will only be modified, could you add documentation to all public methods? Thanks!

**Comment by afchung** [comment]
@anupam128 I'm done with another pass. Thanks!

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs (hunk 18 above)
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: This class will be replaced in the sense that it will be merged with YarnREEFClient once we reach feature parity for this stream of work items.

**Comment by afchung** [comment]
@anupam128 Seems like there are test failures, could you take another look?

**Comment by anupam128** [comment]
@afchung thanks for pointing out. Fixed it.

**Comment by markusweimer** [comment]
@afchung Will you take this across the finish line? I'm +1 on merging this.

**Comment by afchung** [comment]
Looks good to me, I will test and merge.

**Comment by afchung** [comment]
@anupam128 I'm still having one failed test:
```
Test Name:	LegacyJobResourceUploaderCanInstantiateWithDefaultBindings
Test FullName:	Org.Apache.REEF.Client.Tests.LegacyJobResourceUploaderTests.LegacyJobResourceUploaderCanInstantiateWithDefaultBindings
Test Source:	c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client.Tests\LegacyJobResourceUploaderTests.cs : line 40
Test Outcome:	Failed
Test Duration:	0:00:00.1116202

Result Message:	
Test method Org.Apache.REEF.Client.Tests.LegacyJobResourceUploaderTests.LegacyJobResourceUploaderCanInstantiateWithDefaultBindings threw exception: 
Org.Apache.REEF.Tang.Exceptions.InjectionException: Could not invoke constructor: new Org.Apache.REEF.Client.Yarn.LegacyJobResourceUploader(Org.Apache.REEF.Client.Common.IJavaClientLauncher = new Org.Apache.REEF.Client.Common.JavaClientLauncher(), Org.Apache.REEF.Client.Common.IResourceArchiveFileGenerator = new Org.Apache.REEF.Client.Common.ResourceArchiveFileGenerator(new Org.Apache.REEF.Common.Files.REEFFileNames()), Org.Apache.REEF.Client.Common.IFile = new Org.Apache.REEF.Client.Common.DotNetFile(), Org.Apache.REEF.Client.Yarn.IYarnCommandLineEnvironment = new Org.Apache.REEF.Client.Yarn.YarnCommandLineEnvironment()) ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.IO.FileNotFoundException: HADOOP_HOME isn't set.
Result StackTrace:	
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Utilities\Diagnostics\Exceptions.cs:line 55
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Utilities\Diagnostics\Exceptions.cs:line 73
   at Org.Apache.REEF.Client.Yarn.YarnCommandLineEnvironment.GetHadoopHomePath() in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client\YARN\YarnCommandLineEnvironment.cs:line 61
   at Org.Apache.REEF.Client.Yarn.YarnCommandLineEnvironment.GetYarnCommandPath() in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client\YARN\YarnCommandLineEnvironment.cs:line 82
   at Org.Apache.REEF.Client.Yarn.YarnCommandLineEnvironment.Yarn(String[] arguments) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client\YARN\YarnCommandLineEnvironment.cs:line 99
   at Org.Apache.REEF.Client.Yarn.YarnCommandLineEnvironment.GetYarnClasspathList() in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client\YARN\YarnCommandLineEnvironment.cs:line 48
   at Org.Apache.REEF.Client.Yarn.LegacyJobResourceUploader..ctor(IJavaClientLauncher javaLauncher, IResourceArchiveFileGenerator resourceArchiveFileGenerator, IFile file, IYarnCommandLineEnvironment yarn) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client\YARN\LegacyJobResourceUploader.cs:line 56
 --- End of inner exception stack trace ---
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.ConstructorInfo.Invoke(Object[] parameters)
   at Org.Apache.REEF.Tang.Implementations.InjectionPlan.InjectorImpl.InjectFromPlan(InjectionPlan plan) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Tang\Implementations\InjectionPlan\InjectorImpl.cs:line 164
 --- End of inner exception stack trace ---
    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Utilities\Diagnostics\Exceptions.cs:line 55
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Utilities\Diagnostics\Exceptions.cs:line 73
   at Org.Apache.REEF.Tang.Implementations.InjectionPlan.InjectorImpl.InjectFromPlan(InjectionPlan plan) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Tang\Implementations\InjectionPlan\InjectorImpl.cs:line 188
   at Org.Apache.REEF.Tang.Implementations.InjectionPlan.InjectorImpl.GetInstance(INode node) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Tang\Implementations\InjectionPlan\InjectorImpl.cs:line 781
   at Org.Apache.REEF.Tang.Implementations.InjectionPlan.InjectorImpl.GetInstance(String clazz) in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Tang\Implementations\InjectionPlan\InjectorImpl.cs:line 797
   at Org.Apache.REEF.Tang.Implementations.InjectionPlan.InjectorImpl.GetInstance[T]() in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Tang\Implementations\InjectionPlan\InjectorImpl.cs:line 764
   at Org.Apache.REEF.Client.Tests.LegacyJobResourceUploaderTests.LegacyJobResourceUploaderCanInstantiateWithDefaultBindings() in c:\Users\anchung\Documents\GitHub\reef-2\lang\cs\Org.Apache.REEF.Client.Tests\LegacyJobResourceUploaderTests.cs:line 41
```

It seems like this test assumes the existence of a local YARN runtime. Could you fix it? Thanks!
